### PR TITLE
[Merged by Bors] - feat(Data/Fin): Add lemmas `Fin.odd_add_one_iff_even` and `Fin.even_add_one_iff_odd`

### PR DESCRIPTION
**Pull request by IvanRenison**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/Fin/Parity.lean [hunk 1]
```diff
@@ -94,4 +94,18 @@ lemma not_even_iff_odd_of_even {n : ℕ} [NeZero n] (hn : Even n) {k : Fin n} :
   rw [even_iff_of_even hn, odd_iff_of_even hn]
   exact Nat.not_even_iff_odd
 
+lemma odd_add_one_iff_even {n : ℕ} [NeZero n] {k : Fin n} : Odd (k + 1) ↔ Even k := by
+  refine ⟨fun ⟨k, hk⟩ ↦ ?_, Even.add_one⟩
+  have h := congrFun (congrArg HSub.hSub hk) 1
+  rw [add_sub_cancel_right, add_sub_cancel_right] at h
+  rw [h]
+  exact even_two_mul k
```
Comment (soraros) [inline Mathlib/Data/Fin/Parity.lean:102]:
```suggestion
  exact add_right_cancel hk ▸ even_two_mul k
```

File: Mathlib/Data/Fin/Parity.lean [hunk 2]
```diff
@@ -94,4 +94,15 @@ lemma not_even_iff_odd_of_even {n : ℕ} [NeZero n] (hn : Even n) {k : Fin n} :
...
 
+lemma odd_add_one_iff_even {n : ℕ} [NeZero n] {k : Fin n} : Odd (k + 1) ↔ Even k := by
+  refine ⟨fun ⟨k, hk⟩ ↦ ?_, Even.add_one⟩
+  exact add_right_cancel hk ▸ even_two_mul k
+
+lemma even_add_one_iff_odd {n : ℕ} [NeZero n] {k : Fin n} : Even (k + 1) ↔ Odd k := by
+  refine ⟨fun ⟨k, hk⟩ ↦ ?_, Odd.add_one⟩
+  have h := congrFun (congrArg HSub.hSub hk) 1
+  rw [add_sub_cancel_right] at h
+  rw [h]
```
Comment (soraros) [inline Mathlib/Data/Fin/Parity.lean:105]:
```suggestion
  rw [eq_sub_iff_add_eq.mpr hk]
```

Comment (IvanRenison) [reply Mathlib/Data/Fin/Parity.lean]: I did something similar but with the same style that in `odd_add_one_iff_even`, is okay?

Comment (soraros) [reply Mathlib/Data/Fin/Parity.lean]: Up to you I guess, I'm a newbie too.

Comment (IvanRenison) [reply Mathlib/Data/Fin/Parity.lean]: Okey, thanks

File: Mathlib/Data/Fin/Parity.lean [hunk 3]
```diff
@@ -94,4 +94,12 @@ lemma not_even_iff_odd_of_even {n : ℕ} [NeZero n] (hn : Even n) {k : Fin n} :
   rw [even_iff_of_even hn, odd_iff_of_even hn]
   exact Nat.not_even_iff_odd
 
+lemma odd_add_one_iff_even {n : ℕ} [NeZero n] {k : Fin n} : Odd (k + 1) ↔ Even k := by
+  refine ⟨fun ⟨k, hk⟩ ↦ ?_, Even.add_one⟩
+  exact add_right_cancel hk ▸ even_two_mul k
+
+lemma even_add_one_iff_odd {n : ℕ} [NeZero n] {k : Fin n} : Even (k + 1) ↔ Odd k := by
+  refine ⟨fun ⟨k, hk⟩ ↦ ?_, Odd.add_one⟩
+  exact eq_sub_iff_add_eq.mpr hk ▸ (even_add_self k).sub_odd odd_one
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/Fin/Parity.lean:103]: If you're going to golf to the point of using `▸`, you might as well go all the way:
```suggestion
lemma odd_add_one_iff_even {n : ℕ} [NeZero n] {k : Fin n} : Odd (k + 1) ↔ Even k :=
  ⟨fun ⟨k, hk⟩ ↦ add_right_cancel hk ▸ even_two_mul k, Even.add_one⟩

lemma even_add_one_iff_odd {n : ℕ} [NeZero n] {k : Fin n} : Even (k + 1) ↔ Odd k :=
  ⟨fun ⟨k, hk⟩ ↦ eq_sub_iff_add_eq.mpr hk ▸ (even_add_self k).sub_odd odd_one, Odd.add_one⟩
```

**Review by Ruben-VandeVelde — APPROVED** [review]

**Comment by kim-em** [comment]
bors merge

**Pull request closed without merging**